### PR TITLE
feat(connector): [Bluesnap] add cards 3DS support 

### PR DESCRIPTION
**Pull request by SamraatBansal**
## Type of Change

- [ ] Bugfix
- [X] New feature
- [ ] Enhancement
- [ ] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Updating support for [Bluesnap](https://developers.bluesnap.com/v8976-JSON/reference/bluesnap-payment-api-json) payment gateway
Payment Method: card
Flow: [3DS](https://developers.bluesnap.com/reference/3-d-secure-for-api#3-d-secure-with-plain-text-cards)

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [ ] This PR modifies application configuration/environment variables

## Motivation and Context

3DS support for cards via Bluesnap is added, it has involvement of [Bluesnap's SDK](https://developers.bluesnap.com/reference/3-d-secure-for-api#section-step-2-add-the-bluesnap-javascript-file-to-your-checkout-form)

In this if the Customer is not present with the connector, we call their [Vaulted Shopper](https://developers.bluesnap.com/v8976-JSON/reference/create-vaulted-shopper) API.
In case of 3DS we create a payments-field token and then transfer the control to SDK to do all validations and call.
Later we use `Complete` flow to handle the redirect response.

## How did you test it?

Did 3DS tests via Postman for all [card scenario](https://support.bluesnap.com/docs/3d-secure#test-3-d-secure-2)
and ran sanity for assurance.

Screenshot 2023-05-05 at 7 05 24 PM

## Checklist

- [X] I formatted the code `cargo +nightly fmt --all`
- [X] I addressed lints thrown by `cargo clippy`
- [X] I reviewed submitted code
- [X] I added unit tests for my changes where possible
- [ ] I added a [CHANGELOG](/CHANGELOG.md) entry if applicable

### CODE REVIEW

File: crates/router/src/connector/bluesnap.rs [hunk 1]
```diff
@@ -462,16 +571,129 @@ impl ConnectorIntegration<api::Authorize, types::PaymentsAuthorizeData, types::P
...
+        req: &types::PaymentsCompleteAuthorizeRouterData,
+        connectors: &settings::Connectors,
+    ) -> CustomResult<Option<services::Request>, errors::ConnectorError> {
+        Ok(Some(
+            services::RequestBuilder::new()
+                .method(services::Method::Post)
+                .url(&types::PaymentsCompleteAuthorizeType::get_url(
+                    self, req, connectors,
+                )?)
+                .headers(types::PaymentsCompleteAuthorizeType::get_headers(
```
Comment (jagan-jaya) [inline crates/router/src/connector/bluesnap.rs:674]:
```suggestion
                .attach_default_headers()
                .headers(types::PaymentsCompleteAuthorizeType::get_headers(
```

File: crates/router/src/connector/bluesnap/transformers.rs [hunk 2]
```diff
@@ -116,10 +127,82 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for BluesnapPaymentsRequest {
...
+}
+
+impl TryFrom<&types::PaymentsCompleteAuthorizeRouterData> for BluesnapPaymentsRequest {
+    type Error = error_stack::Report<errors::ConnectorError>;
+    fn try_from(item: &types::PaymentsCompleteAuthorizeRouterData) -> Result<Self, Self::Error> {
+        let redirection_response: BluesnapRedirectionResponse = item
+            .request
+            .payload
+            .clone()
+            .ok_or(errors::ConnectorError::MissingRequiredField {
```
Comment (jagan-jaya) [inline crates/router/src/connector/bluesnap/transformers.rs:142]: add a new error for `MissingPayloadField{fieldname}`, MissingRequiredField is reserved for customer missing an required field. this is not customer input.

File: crates/router/src/connector/bluesnap/transformers.rs [hunk 3]
```diff
@@ -116,10 +127,82 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for BluesnapPaymentsRequest {
...
+        {
+            PaymentMethodDetails::CreditCard(Card {
+                card_number: ccard.card_number,
+                expiration_month: ccard.card_exp_month.clone(),
+                expiration_year: ccard.card_exp_year.clone(),
+                security_code: ccard.card_cvc,
+            })
+        } else {
+            Err(errors::ConnectorError::MissingRequiredField {
+                field_name: "PaymentMethod",
```
Comment (jagan-jaya) [inline crates/router/src/connector/bluesnap/transformers.rs:168]: add a new error for `MissingPayloadField{fieldname}`, MissingRequiredField is reserved for customer missing an required field. this is not customer input.

File: crates/router/src/connector/bluesnap/transformers.rs [hunk 4]
```diff
@@ -116,10 +127,82 @@ impl TryFrom<&types::PaymentsAuthorizeRouterData> for BluesnapPaymentsRequest {
...
+        };
+        Ok(Self {
+            amount: utils::to_currency_base_unit(item.request.amount, item.request.currency)?,
+            payment_method,
+            currency: item.request.currency,
+            card_transaction_type: auth_mode,
+            three_d_secure: Some(BluesnapThreeDSecureInfo {
+                three_d_secure_reference_id: redirection_result
+                    .three_d_secure
+                    .ok_or(errors::ConnectorError::MissingRequiredField {
```
Comment (jagan-jaya) [inline crates/router/src/connector/bluesnap/transformers.rs:179]: add a new error for `MissingPayloadField{fieldname}`, MissingRequiredField is reserved for customer missing an required field. this is not customer input.

File: crates/router/src/connector/bluesnap.rs [hunk 5]
```diff
@@ -760,3 +982,31 @@ impl api::IncomingWebhook for Bluesnap {
...
+
+impl services::ConnectorRedirectResponse for Bluesnap {
+    fn get_flow_type(
+        &self,
+        _query_params: &str,
+        json_payload: Option<serde_json::Value>,
+        _action: services::PaymentAction,
+    ) -> CustomResult<payments::CallConnectorAction, errors::ConnectorError> {
+        let redirection_response: bluesnap::BluesnapRedirectionResponse = json_payload
+            .ok_or(errors::ConnectorError::MissingRequiredField {
```
Comment (jagan-jaya) [inline crates/router/src/connector/bluesnap.rs:994]: add a new error for `MissingPayloadField{fieldname}`, MissingRequiredField is reserved for customer missing an required field. this is not customer input.

File: crates/router/src/utils.rs [hunk 6]
```diff
@@ -119,3 +119,32 @@ impl<E> ConnectorResponseExt
             })
     }
 }
+
+/// Convert the amount to its base denomination based on Currency and return String
+pub fn to_currency_base_unit(
+    amount: i64,
+    currency: storage_models::enums::Currency,
+) -> Result<String, error_stack::Report<errors::ConnectorError>> {
```
Comment (jagan-jaya) [inline crates/router/src/utils.rs:127]: ConnectorError here?

File: crates/router/src/utils.rs [hunk 7]
```diff
@@ -119,3 +119,32 @@ impl<E> ConnectorResponseExt
...
+    let amount_f64 = to_currency_base_unit_asf64(amount, currency)?;
+    Ok(format!("{amount_f64:.2}"))
+}
+
+/// Convert the amount to its base denomination based on Currency and return f64
+pub fn to_currency_base_unit_asf64(
+    amount: i64,
+    currency: storage_models::enums::Currency,
+) -> Result<f64, error_stack::Report<errors::ConnectorError>> {
```
Comment (jagan-jaya) [inline crates/router/src/utils.rs:136]: ConnectorError here? change it to Validation error

**Review by jagan-jaya — CHANGES_REQUESTED** [review]

File: config/development.toml [hunk 8]
```diff
@@ -179,33 +179,33 @@ ideal = { country = "NL", currency = "EUR" }
 
 [pm_filters.braintree]
 paypal = { currency = "AUD,BRL,CAD,CNY,CZK,DKK,EUR,HKD,HUF,ILS,JPY,MYR,MXN,TWD,NZD,NOK,PHP,PLN,GBP,RUB,SGD,SEK,CHF,THB,USD" }
-credit = { not_available_flows = {capture_method="manual"} }
-debit =  { not_available_flows = {capture_method="manual"} }
+credit = { not_available_flows = { capture_method = "manual" } }
```
Comment (jagan-jaya) [inline config/development.toml:182]: are these formatting intentional?

Comment (SamraatBansal) [reply config/development.toml:182]: Cargo fmt, did this formatting, not done intentionally

File: crates/router/src/connector/bluesnap/transformers.rs (hunk 3 above)
Comment (SamraatBansal) [reply crates/router/src/connector/bluesnap/transformers.rs:168]: This is the `payment_method` field, How should it be handled? Its not a connector error

Comment (jagan-jaya) [reply crates/router/src/connector/bluesnap/transformers.rs:166]: No this block requires a ConnectorError error only, we can't throw other errors so, add a new error MissingPayloadField{fieldname} in ConnectorError and use it

File: crates/redis_interface/src/lib.rs [hunk 9]
```diff
@@ -155,13 +155,6 @@ impl RedisConnectionPool {
     }
 }
 
-impl Drop for RedisConnectionPool {
```
Comment (jarnura) [inline crates/redis_interface/src/lib.rs:158]: Why this change is reverted?

**Review by jarnura — CHANGES_REQUESTED** [review]

**Review by jarnura — DISMISSED** [review]

File: crates/router/src/connector/bluesnap.rs [hunk 10]
```diff
@@ -462,16 +571,130 @@ impl ConnectorIntegration<api::Authorize, types::PaymentsAuthorizeData, types::P
         data: &types::PaymentsAuthorizeRouterData,
         res: Response,
     ) -> CustomResult<types::PaymentsAuthorizeRouterData, errors::ConnectorError> {
+        match (data.is_three_ds(), res.headers) {
+            (true, Some(headers)) => {
+                let location = connector_utils::get_http_header("Location", &headers)?;
+                let payment_fields_token = url::Url::parse(location)
+                    .into_report()
+                    .change_context(errors::ConnectorError::URLParsingFailed {
```
Comment (ArjunKarthik) [inline crates/router/src/connector/bluesnap.rs:579]: Instead of parsing this to URL we can parse the string itself. And there is no need for new error type "URLParsingFailed"

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/router/src/connector/bluesnap.rs [hunk 11]
```diff
@@ -128,6 +135,100 @@ impl ConnectorIntegration<api::Verify, types::VerifyRequestData, types::Payments
...
+            .response
+            .parse_struct("BluesnapCustomerResponse")
+            .change_context(errors::ConnectorError::ResponseDeserializationFailed)?;
+
+        types::RouterData::try_from(types::ResponseRouterData {
+            response,
+            data: data.clone(),
+            http_code: res.status_code,
+        })
+        .change_context(errors::ConnectorError::ResponseHandlingFailed)
```
Comment (ArjunKarthik) [inline crates/router/src/connector/bluesnap.rs:221]: change_context not required

**Review by ArjunKarthik — CHANGES_REQUESTED** [review]

File: crates/redis_interface/src/lib.rs (hunk 9 above)
Comment (SamraatBansal) [reply crates/redis_interface/src/lib.rs]: There was a breaking change, Reverted to test, No Re-reverted after hotfix in #1063

**Review by ArjunKarthik — APPROVED** [review]
@SamraatBansal reviewed connector code and it looks fine to me

**Review by jagan-jaya — DISMISSED** [review]

**Review by jagan-jaya — APPROVED** [review]

File: crates/router/src/services/api.rs [hunk 12]
```diff
@@ -760,6 +785,72 @@ pub fn build_redirection_form(form: &RedirectForm) -> maud::Markup {
...
+                card_details = format!(
+                    "var newCard={{ccNumber: \"{}\",cvv: \"{}\",expDate: \"{}/{}\",amount: {},currency: \"{}\"}};",
+                    ccard.card_number.expose(),
+                    ccard.card_cvc.expose(),
+                    ccard.card_exp_month.clone().expose(),
+                    ccard.card_exp_year.expose(),
+                    amount,
+                    currency
+                );
+            };
```
Comment (dracarys18) [inline crates/router/src/services/api.rs:802]:
```suggestion
            let mut card_details = if let Some(api::PaymentMethodData::Card(ccard)) = payment_method_data {
                card_details = format!(
                    "var newCard={{ccNumber: \"{}\",cvv: \"{}\",expDate: \"{}/{}\",amount: {},currency: \"{}\"}};",
                    ccard.card_number.expose(),
                    ccard.card_cvc.expose(),
                    ccard.card_exp_month.clone().expose(),
                    ccard.card_exp_year.expose(),
                    amount,
                    currency
                );
            } else {"".to_string()};
```

**Review by jarnura — APPROVED** [review]

**Review by jagan-jaya — APPROVED** [review]
